### DISCUSSION TITLE
Feature Announcements Thread

### DISCUSSION THREAD

**Discussion by mcmonkey4eva** [Stability-AI/StableSwarmUI#11] [Announcements]
This is a dedicated thread you can follow where new features will be announced.

While the project is in alpha, things will likely move fast and there will be posts here somewhat often. After alpha is done, I'll probably open a Releases thread for a slower roundup.

# Migration Notice

As of 2024/06/21 StableSwarmUI will no longer be maintained under Stability AI.

The original developer will be maintaining an independent version of this project as [mcmonkeyprojects/SwarmUI](https://github.com/mcmonkeyprojects/SwarmUI)

Windows users can migrate to the new independent repo by simply updating and then running migrate-windows.bat

For Linux, Mac, or manual Windows: open a terminal in your Swarm folder then run git remote set-url origin https://github.com/mcmonkeyprojects/SwarmUI

See full migration guide here: https://github.com/mcmonkeyprojects/SwarmUI/discussions/2

**Comment by mcmonkey4eva** [comment]
Key feature update: **refiner native integration!** If you use ComfyUI backend, the refiner stage is now readily supported.

There's a checkbox to download it while you install, and:
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/db6a5422-32ac-42a9-b3d1-83cce4903e73)
You can choose whether to refine, if so with which model, and which method (StepSwap or PostApply), how strongly to refine, and optionally whether to upscale the image (and if so, how).

You can also use mid-run upscaling on its own without swapping the model (aka highres fix)

(This is for the easy-mode without automatic workflow generation, if you use custom workflows this doesn't change, you could already do refiners that way)

**Comment by mcmonkey4eva** [comment]
When you use a custom Comfy workflow, it now is able to automatically recognize the list of valid inputs - so for example if you have a LoRA loader node, it will show your list of lora files
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/676576b8-d87c-406a-a2b1-c6cf9b4e05ce)
this also applies to the Grid Generator tool, wherein the `Fill` button now works to fill the appropriate values

**Comment by mcmonkey4eva** [comment]
A lot of work has been done on custom comfy workflows lately.

Most notably, you can now title a Primitive input with the format `SwarmUI: Prompt` to tell swarm to automatically recognize that input as using the default Prompt parameter. You can also do this for any other parameters, as long as the name is correct. (and caps/spacing/etc is ignored, so eg `swarmui:seed`, `Swarmui: comfyui_sampler`, etc. all will work)
![image-2](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/365fad5f-3f52-4620-935e-02843c6bacb0)

I use prompt in this exactly as that's commonly broken by using the advanced encoder, but naming the primitive to get it back is very useful for having easy compatibility with presets.

**Comment by mcmonkey4eva** [comment]
Image History, Presets, and Models (the 3 Browsable parts) now have a 'Depth' input so you can view x many folders of depth, and a Filter text box to only show results that match the filter text (eg you can search "SDXL" in the models area to only show SDXL models) (filter checks name/description/type/etc for models, name/metadata for images, etc - case insensitive)

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/6c880bed-6e7e-4994-a79a-fc1ac9ff0c92)

Also it has lazy-load on images, so if you view deep image history it won't die trying to load the images, it loads as you scroll.

Also there's a cap of 1000 results while deep-viewing images to avoid issues, this will be configurable in the future.

**Comment by mcmonkey4eva** [comment]
There's now a working interface for LoRAs and VAEs (for generated workflows, doesn't do anything for custom workflows).

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/3500757e-cb1a-4ac2-b0e1-7bec25769d42)

LoRAs selected show up in the status bar, with a strength adjustment input.

Just as with the models page, any lora with proper metadata on it will render its metadata
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/2bacf025-d13d-4730-8795-3b4816d5c32d)

And you can edit the metadata freely via the lil menu button.


(Also a browser for Embeddings that at the moment doesn't do anything functional.)

**Comment by mcmonkey4eva** [comment]
Custom Comfy Workflows can be saved, loaded, and compared in grids now! (You can also delete them from the Load menu)

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/087e34b2-a852-4d64-a34f-5b563f111dbe)

**Comment by mcmonkey4eva** [comment]
There's now a (basic initial) ControlNet option on the main generate tab
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/71dc36bb-1a09-4927-98e7-fed60e2d741b)

**Comment by mcmonkey4eva** [comment]
You can now quickshut the leftside parameter view with a button press and get a simplified/focused prompt-only view:

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/a4e58b1e-24da-4390-b40b-5254bf2245c0)

Just you, your image, and your prompt. Hit enter to quicksend the prompt. But all the fancy params are still accessible by popping open the left bar at any time.

**Comment by mcmonkey4eva** [comment]
Initial ReVision native support:

![ReVision](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/d998ef38-eb36-4f1f-9f36-e4127d721584)

Just slap some images to the prompt box and it'll figure it out for you. Only compatible with Comfy backend + default workflow generator. Will autodownload clip_vision_g for you if you lack it and you're using comfy-self-start and/or at least have your model dir pointed correctly.

**Comment by mcmonkey4eva** [comment]
There's a css/format rework in-progress. You can `git checkout cec01a12c8533d1fc7934cfa9abd99fe0f52a22f` to manually go back to before this if it causes any issues for you.

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/d40080f3-ca14-47ee-bf31-fd1e57a0b5fe)

Notably parameters now are less space-heavy, and Prompt/Generate are now in the main center area, so you can just close the parameter list and focus on prompting when you're not in the mood for fiddling settings.

This is still a work-in-progress, there's more changes planned before this is done.

**Comment by mcmonkey4eva** [comment]
You can now right-click the generate button to enable "Generate Forever", which will queue up a new generate each time the last one finishes, so you can see the results of your changes as you adjust things. Works best with fast models / low step counts / static seed.

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/a362bc8d-49f6-4089-95f3-8a8a1618fcee)

You can also right-click the interrupt `x` button to get "Interrupt All Sessions", so if you accidentally close a window that has a long queue and want it to just stop, you can use that button in a new window to stop it.

Interrupt is also available under Quick Tools.

**Comment by mcmonkey4eva** [comment]
There's now a tab under User where you can muck with some of the raw parameter configuration (eg number view type / min/max, or example values as are filled by gridgen, or etc)
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/fafe2131-0dd3-4eae-a7a0-afbc50a20d28)

**Comment by mcmonkey4eva** [comment]
The Grid Generator tool now has search + tab complete as you type.
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/0fd75637-3b1a-4610-a037-7d11526e0b67)
You can click a suggestion with your mouse, or use Arrow key up/down to navigate and then hit Tab to use it.

**Comment by mcmonkey4eva** [comment]
I published a convenient download for the SDXL official preset pack in a Swarm-ready format, including icons and all!
Find that doc here: https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Presets.md

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/a212d8bc-7d66-4ec4-8af9-8353c47e7c02)

**Comment by mcmonkey4eva** [comment]
In my time circling SD-related support channels, I've answered user questions many times about "how does CLIP tokenize this text?" / "does CLIP recognize this symbol?" / "Is this even a word CLIP knows?" / etc. So... Swarm has a utility to answer those questions easily now:
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/66d9c2d6-6158-4cfc-b904-107955aa1327)

It's under the new "Utilities" tab, and is a relatively basic interface, but works well.

Also! When generating images, any prompt input now comes with a CLIP Token counter that shows how many tokens your current prompt has, and tells you when you hit the CLIP 75-token break line (just like what Auto WebUI has)
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/c9938d2f-1f41-4d56-bb3e-d6181c27d78c)

**Comment by mcmonkey4eva** [comment]
Have a lot of images you want to turn into different images? There's a tool for that now! It is very barebones and built-for-purpose, not fancy at all, but if you have the need, this tool is the answer.
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/aa3260dc-fd5f-4b2a-8fab-d46a559916ed)
(this also is serving as a test case for extensions that break the normal control pattern to start looking into how to better handle that. The "paste a link to a folder here" setup is _not_ how swarm normally does things, and the goal is to experiment with (A) how well can extensions do things the lazy way like this and still work, and (B) how can we convert things that necessarily have awkward inputs like a folder-load of images into a clean interface?)

---

ps there's also now a lil toggler at the top right over the batch view:
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/e103432a-b094-4e6f-8a8c-d42549bf3fe0)
so if you want the batch list to not clear when you generate more things, that button does that. It persists in browser local storage, so as long as you're not using ngrok/cloudflared the choice should persist.

**Comment by mcmonkey4eva** [comment]
New backend type: `Swarm-API-Backend`. This is a backend that connects to _another instance of StableSawrmUI_. This is especially intended for LAN/network connections - you can just install & run Swarm on the other machine, and then have your main swarm instance connect to the remote swarm instance. You don't have to worry about configuring backend stuff manually!
It is, in theory, implicitly compatible with, er, everything! It can even redirect a Comfy Workflow Editor wrapper page across network cleanly.

Note that you still have to make sure your model list is synchronized between your local machine and the remote instance.

**Comment by mcmonkey4eva** [comment]
LoRAs and ControlNets now know if they're compatible with your currently selected model or not - if they don't match, they fade out and get sorted to the bottom.
So eg if you're running SDXL, any SDv1 LoRAs will be out-of-the-way when you're looking at your lora options.

(This theoretically applies to embeddings too but in practice most embeddings out there are in pickle `.pt` files so they don't have metadata - I intend in the future to add an easy utility to batch convert your model files to solve that)

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/d3a56c6f-3004-49f3-bce4-d6edb7a63475)

**Comment by mcmonkey4eva** [comment]
_Prompt syntax_ support is now in the works. Including eg `<embedding:embedname>`, `<lora:filename:1>`, `<random:a, b, c>`, `<preset:presetname>`

For example `a photo of a <random:orange, green, blue, purple, glass, shiny silver, black> cat` with images=10 gets you:
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/71e941b0-1322-4db0-a8d2-a88cf566b5f0)

I wanted to make an intentionally consistent syntax (ie it is always `<prefix:data>`) as opposed to eg Auto WebUI where embeds are just `filename`, loras are `<lora:filename:1>` and random requires an extension and is `{a|b|c}`.

I intend to add tools to highlight/tab-complete/etc. to make it very easy & clear to use - for now it's at least documented @ https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Basic%20Usage.md#more-than-text-playing-with-prompts

Also you can select text and control+up/down arrow to adjust prompt weighting.

**Comment by mcmonkey4eva** [comment]
New (Comfy-only) auto-segmentation-detailer! aka, "ADetailer"++

In short: by adding the new `<segment:x>` syntax to your prompt, you can trigger an automatic cropped refinement of any part of your imagest that you want. Commonly this might be used for faces.

For example: `wide shot of a beautiful woman in a green sweater, smiling, full body, standing
<segment:face>a beautiful hd photo of a woman's face, smiling, bright eyes, best quality, highres, film grain`
This will generate a photo of a woman, then it will zoom in on her face, and edit the face.

# Details

This is automatically compatible with refiners/etc (if you use a refiner model, that will be used for the segment edit instead of your base)

The segmentation uses [CLIP Seg](https://huggingface.co/docs/transformers/model_doc/clipseg) to automatically detect the part of the image matching your description using CLIP. Simple words like `face` do fine, you can get more complex if you want.

If the segmentation isn't working as well as you need out of the box, you can do `<segment:text,creatitivity,threshold>` eg `<segment:face,0.7,0.5>` where:
- `creativity` is Init Image Creativity aka Denoising Strength - `1.0` means totally replace the image, `0.0` means no change. Default is `0.6`.
- `threshold` is CLIP Seg's threshold. ie the minimum match ratio, where `0.0` means capture basically the whole image, and `1.0` means capture nothing. Default is `0.5`.

You can do multiple segments in one prompt. For convenient organization, use Shift+Enter to add newlines between segment lines.

Segment requests are only accepted in the main Prompt box, _however_ you may add segments in the negative box will the same match text (so eg if you have positive `<segment:face>` you can also use negative `<segment:face>` - the text must be identical, so copy/paste between inputs to be safe) to change the negative prompt. If you don't have a negative segment prompt, your standard negative prompt will be used for the segment.

# Examples

```
a photo of a man petting a cat in a park
<segment:cat> hd photo of a cat, beautiful, film grain, furry masterpiece
<segment:man's face> a close up photo of a man's face, beautiful, film grain, masterpiece
```
![2151-a photo of a man petting a cat in a park-OfficialStableDiffusionsd_xl_base_10s-732224291](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/95973277-353b-471d-9006-b5133fa3adc5)

```
a photo of a cat walking in a park
<segment:cat's head,0.8,0.5>a photo of a dog's head
```
![2236-a photo of a cat walking in a parksegm-OfficialStableDiffusionsd_xl_base_10s-1619443027-2](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/843a417f-1dd2-4b99-af57-8f5417a60226)

**Comment by mcmonkey4eva** [comment]
If you use custom Comfy workflows with video tools like [AnimateDiff](https://github.com/Kosinkadink/ComfyUI-AnimateDiff-Evolved) or [VideoHelper](https://github.com/Kosinkadink/ComfyUI-VideoHelperSuite), swarm will now understand animated `.gif` file output completely, and also (partially) supports `.webm`/`.mp4` files!
This means you can use all the feature of Swarm with your animation/video projects (including multi-GPU, Grids, etc.)

You'll still have you figure out a good workflow for it though, that's not automated away yet. (Hopefully you can make one better than the one I used to test and make the attached gif, heh)

![badanim](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/1a22638f-e25d-4ede-b4dd-bf42cfb0452e)

**Comment by mcmonkey4eva** [comment]
With thanks to @surfingdoggo in #88 - Swarm is now dockerable! See info in readme - https://github.com/Stability-AI/StableSwarmUI#running-with-docker

**Comment by mcmonkey4eva** [comment]
New Utility: Pickle To Safetensors bulk converter!
Finally those of us with giant model folders full of random old scrap we downloaded months ago can clean that folder up the easy way.
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/05909b3f-8858-4555-9122-8cf2e424cdc6)

This can convert pickles in any of the main model categories to modern safetensors files.

The FP16 conversion (and ema/etc. scrap discard) will also help you save space (eg an SDv1 pickle that's 7GiB or 4GiB will be shrunk down to 2GiB with effectively no quality loss).

Also, as a bonus, when you convert TI Embeddings (the most common pickles still found out there), it will insert the model's original name and training target in the model's metadata (because for some reason these are both stored in embed pickles)
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/9131a406-e73b-4231-b2af-df67a55ddeb2)

**Comment by mcmonkey4eva** [comment]
ComfyUI Workflow Editor tab can now use multiple backends simultaneously! At the top left under `MultiGPU` selector choose `All`, then create a workflow where the output nodes (PreviewImage/SaveImage) have different colors (use 1 unique color per backend). Each output's sub-graph will be split to a different comfy backend.

(This naturally only works if you have multiple comfy-based backends)

This video shows 1 prompt, 2 models, 8 seeds:

https://github.com/Stability-AI/StableSwarmUI/assets/4000772/3891d5f0-7518-4b42-93cc-92c978ef5514

**Comment by mcmonkey4eva** [comment]
You can now set up generations how you want on the Generate tab with the easy simple UI there, and then click "Import" on the comfy tab and view the raw comfy nodegraph guts of it!

![autoimportcomfy](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/e93c7fe3-7ad3-4a9b-8f7f-463d9f57ec62)

**Comment by mcmonkey4eva** [comment]
Wildcards! You can now create wildcard files and use them. You can manage the entirely in the UI, -or- just go store some `.txt` files into `Swarm/Data/Wildcards` (or configure the path in server settings).

Use in a prompt like `<wildcard:animals>`, or just click the card in the UI.

Wildcard files are simple text - one entry per line.
Wildcards can contain prompt-syntax (eg another wildcard inside the wildcard, or a preset, or whatever you want)
Use `#` to comment a line (not count in the options list).

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/b60a8d0b-1e47-4cac-93ff-179c2a40611d)

**Comment by mcmonkey4eva** [comment]
SVD! Want to make some Stable Videos? Swarm's main interface now supports this, via the "Video" parameter group, as long as you have an SVD model installed in your models folder.
Note that this is a first pass of support, and has a rough/raw interface for the moment.

I recommend 16:9 aspect for the text2image pass, as that converts well to the models rather unusal native res of 1024x576, otherwise you'll get clipped off. There are some parameters hidden under "advanced".

You can use an Init Image with Creativity=0 to effectively just do img2video as in the original demo.

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/8f9573c0-140e-43d1-88f0-48cba34dc84a)

![view](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/121d48fc-1ac9-405a-b0b5-767c629a097a)

**Comment by mcmonkey4eva** [comment]
New!: Log viewer in UI

You can now view server Logs, under the Server tab.

You can select which log type to view, eg you can view Debug or raw logs from a backend

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/2298a0a5-10de-4a1d-ae85-d05885d8f060)

This especially makes debugging errors much easier (previously you'd have to fiddle log level settings and restart and etc. Now you just go look at the logs viewer and select to view Debug if you need that, or verbose or a backend or...)

**Comment by mcmonkey4eva** [comment]
New: LoRA Extractor Utility

As it says on the tin: lets you extract LoRAs from your models

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/c9c5f771-b8f3-4e3e-bede-4c8ca158cfaf)

**Comment by mcmonkey4eva** [comment]
New: Tab completion in prompts! Just type a `<` and you will be suggested all available prompt syntax. Most prompt syntax options have further tab completions once selected.

![tab-completey](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/b2899e95-2288-4dbb-8357-f66a9cebab56)

**Comment by mcmonkey4eva** [comment]
New! Automatic welcome messages.

Whenever you first open swarm, it will have a welcome message. This can be one of three categories:

- If you're new to swarm (haven't selected a model), it will tell you to select a model. Once you do that, it will switch to telling you to type a prompt below and hit generate. I hope to expand this with more helpful getting-started tips.
- Recent new features selected at random, (to improve awareness of feature updates)
- Random "did you know" tips for interesting or lesser-known features of swarm worth giving a try

These messages occupy the location where generated images would be, and as such go away as soon as you generate an image or pull one up from history.

You can refresh the page at any time to see a new random welcome message.

There is a server setting available (UI -> `OverrideWelcomeMessage`) to manually set a custom welcome message body if you don't like seeing the generated ones, or if you're sharing a swarm instance and want to show a message to your users.

If you're curious, you can see the full list of defined welcome messages @ https://github.com/Stability-AI/StableSwarmUI/blob/master/src/wwwroot/js/genpage/welcomemessages.js

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/e2197601-189a-4294-a9f7-8cc92428fe3d)

**Comment by mcmonkey4eva** [comment]
Grid generator upgrade!

Tools -> Grid Generator got some fun new tricks recently.

- Has a filled out readme now https://github.com/Stability-AI/StableSwarmUI/blob/master/src/BuiltinExtensions/GridGenerator/README.md

- There's now a dropdown to select between Web Page, Just Images, or Grid Image
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/6b632197-dfea-43b6-9712-ed0fd1b61bb2)
Naturally, web page is the classic full-detail magic webpage it always had, Just Images will bulk-generate images without making a grid out of them, Grid Image will do a simple single 2D giant grid image at the end like that old auto webui script did

- There's now Save/Load config buttons
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/430d6691-6789-40ef-9ff4-25fb6685860d)
These save/load your entire grid setup - output type, folder name (for web page gen), checkbox/settings, and axes/values. This allows you to easily recontinue prior grids on the fly.

- You can now use `SKIP:` (all caps) before a value to skip it. Useful for the above mentioned reusing-an-old-grid concept.

- There's now a lil hint spot mentioning the new tricks with a link to the new readme about the grid generator. Hover your mouse over the hint to have it light up, otherwise it's faded to not be annoying.
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/c90e6cc3-0c24-4882-bdd5-87790ea1fb28)

- There's now live error checking as you type, automatically detecting invalid inputs where possible to highlight them red. Hover your mouse over a red bubble to see why it was rejected.
![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/2d914286-9967-48f5-bf9a-2ff2579493e5)

**Comment by mcmonkey4eva** [comment]
Installer improvements!

- There's now a clean install script for Linux users.
- The installer UI is not a lot nicer
    - no longer shoves a whole license in your face at the start
    - has a bunch to skip all the configuration steps and just use the defaults, since the defaults are best for most users anyway
    - now shows a progress bar as its downloading files at the end.

**Comment by mcmonkey4eva** [comment]
New! Translations! The UI is now translated into several common languages, and many more will be added in the coming few days.

![image](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/1670e4c8-53e6-4a14-a30f-8e0b3312310f)

I bulk generated the initial translations using an AI model from google (madlad400-7b-mt-bt) using a tool I built to quick bulk that task: https://github.com/mcmonkeyprojects/translate-tool

I welcome PRs to improve the quality of the translations, see also https://github.com/Stability-AI/StableSwarmUI/issues/80#issuecomment-1891472595

**Comment by mcmonkey4eva** [comment]
Stable Cascade support! 

See model support doc here for specifics: https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Model%20Support.md#stable-cascade

Simply use the "ComfyUI" / "All In One" format models and save them together in your checkpoints folder https://huggingface.co/stabilityai/stable-cascade/tree/main/comfyui_checkpoints

**Comment by mcmonkey4eva** [comment]
Comfy workflow browser!

Are you a ComfyUI fanatic with hundreds of workflows? Organize them with names/thumbnails/etc. with a quick and friendly UI, directly in Swarm's Comfy tab!
~~*Swap through bowls of spaghetti like someone who's about to get kicked out of the italian all-you-can buffet!*~~

![comfy-workflow-browser](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/79ff6a22-3f9b-4f55-a9b6-c2c075ef9474)

**Comment by mcmonkey4eva** [comment]
New themes: **Modern Dark** and **Modern Light** are now available, with _Modern Dark_ as the new default theme!

![image-3](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/26f904d9-ca65-498c-a1f1-3bc2a4130ee0)

Built by Stability AI designer Eric and pythongosssss



... also, for fun, here's the first design idea Eric made, before we started simplifying things down a bit:
![image-3](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/7982feb5-d333-4638-811c-ce20ccee3d6a)
(What do you think? Should we have gone with the this "retro control panel" aesthetic? Is there something that could be made better? Feel free to poke me in https://discord.gg/stablediffusion `#stableswarmui` channel with opinions, or start a discussion in the Discussions tab here on github)

**Comment by mcmonkey4eva** [comment]
**Prompt editing!** One of the most requested features - a hit from Auto WebUI, long missing in other UIs, now available in Swarm with comfy backend!
These features are a fun way to mix distinct concepts in one generation.

For alternating words, use `<alternate:cat, dog>`
For editing prompts between timesteps, use `<fromto[0.5]:cat, dog>`
See docs here: https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Prompt%20Syntax.md
also if you have auto webui syntax for this copypasted, like `[cat|dog]` or `[cat:dog:0.5]`, this will be recognized and work automatically.


![alternatingdemo](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/b7afa326-caa4-441f-8708-a1783c33a9d6)

**Comment by mcmonkey4eva** [comment]
Swarm has a dedicated full discord of its own now! https://discord.gg/q2y38cqjNw

**Comment by mcmonkey4eva** [comment]
Are you a fan of adetailer's YOLOv8 face detection and/or the per-person matching logic?

Now you can use YOLOv8 in Swarm - https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Features/Prompt%20Syntax.md#automatic-segmentation-and-refining

tldr: download a yolov8 model like https://github.com/hben35096/assets/releases/download/yolo8/face_yolov8m-seg_60.pt to `(Swarm)/Models/yolov8`, then in swarm type a prompt like `a photo of a woman standing in the street smiling <segment:yolo-face_yolov8m-seg_60.pt-1> a beautiful woman's smiling face` with this new segment sub-syntax to automatically use YOLOv8 for the segmentation handling.

**Comment by mcmonkey4eva** [comment]
TensorRT support! Thanks to Comfy and nvidia for building the Comfy toolkit for TRT

Simply click the "Create TensorRT Engine" button in the menu for a model
![image-3](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/d455cc35-2291-412d-8aac-6e1fe40247c1)


 you will then get a `.engine` file listed in the same folder (on filesystem this is actually in `(Models)/tensorrt/`)
TRT models are great for static servable models that run faster that normal models do (eg that you'll use over an API or just use frequently) but aren't compatible with all options (eg loras and wotnot are unsupported currently)

Note as well this is brand new / experimental and may have reliability issues

**Comment by mcmonkey4eva** [comment]
New **Model Downloader** Utility tab

![autodownloadmodels](https://github.com/Stability-AI/StableSwarmUI/assets/4000772/ef5b3d58-8fdd-4026-ba00-c8de4dc9c9eb)

Notably supports entering a civitai url and it will download the model + metadata for that model automatically

You can also click "Edit Metadata" on an existing model and give a civitai URL to load model from that url onto an existing model

**Comment by mcmonkey4eva** [comment]
As of 2024/06/21 StableSwarmUI will no longer be maintained under Stability AI.

The original developer will be maintaining an independent version of this project as [mcmonkeyprojects/SwarmUI](https://github.com/mcmonkeyprojects/SwarmUI)

Windows users can migrate to the new independent repo by simply updating and then running migrate-windows.bat

For Linux, Mac, or manual Windows: open a terminal in your Swarm folder then run git remote set-url origin https://github.com/mcmonkeyprojects/SwarmUI

See full migration guide here: https://github.com/mcmonkeyprojects/SwarmUI/discussions/2
